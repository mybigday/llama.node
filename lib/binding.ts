import * as path from "path";

export type ChatMessage = {
  role: string;
  content: string;
};

export type LlamaModelOptions = {
  model: string;
  chat_template?: string;
  embedding?: boolean;
  embd_normalize?: number;
  pooling_type?: "none" | "mean" | "cls" | "last" | "rank";
  n_ctx?: number;
  n_batch?: number;
  n_ubatch?: number;
  n_threads?: number;
  n_gpu_layers?: number;
  flash_attn?: boolean;
  cache_type_k?:
    | "f16"
    | "f32"
    | "q8_0"
    | "q4_0"
    | "q4_1"
    | "iq4_nl"
    | "q5_0"
    | "q5_1";
  cache_type_v?:
    | "f16"
    | "f32"
    | "q8_0"
    | "q4_0"
    | "q4_1"
    | "iq4_nl"
    | "q5_0"
    | "q5_1";
  use_mlock?: boolean;
  use_mmap?: boolean;
  vocab_only?: boolean;
  lora?: string;
  lora_scaled?: number;
  lora_list?: { path: string; scaled: number }[];
};

export type CompletionResponseFormat = {
  type: "text" | "json_object" | "json_schema";
  json_schema?: {
    strict?: boolean;
    schema: object;
  };
  schema?: object; // for json_object type
};

export type LlamaCompletionOptions = {
  messages?: ChatMessage[];
  jinja?: boolean;
  chat_template?: string;
  response_format?: CompletionResponseFormat;
  tools?: object;
  parallel_tool_calls?: boolean;
  tool_choice?: string;
  prompt?: string;
  temperature?: number;
  top_k?: number;
  top_p?: number;
  min_p?: number;
  mirostat?: number;
  mirostat_tau?: number;
  mirostat_eta?: number;
  penalty_last_n?: number;
  penalty_repeat?: number;
  penalty_freq?: number;
  penalty_present?: number;
  typ_p?: number;
  xtc_threshold?: number;
  xtc_probability?: number;
  dry_multiplier?: number;
  dry_base?: number;
  dry_allowed_length?: number;
  dry_penalty_last_n?: number;
  n_predict?: number;
  max_length?: number;
  max_tokens?: number;
  seed?: number;
  stop?: string[];
  grammar?: string;
  grammar_lazy?: boolean;
  grammar_triggers?: { word: string; at_start: boolean }[];
  preserved_tokens?: string[];
};

export type LlamaCompletionResult = {
  text: string;
  tokens_predicted: number;
  tokens_evaluated: number;
  truncated: boolean;
  timings: {
    prompt_n: number;
    prompt_ms: number;
    prompt_per_token_ms: number;
    prompt_per_second: number;
    predicted_n: number;
    predicted_ms: number;
    predicted_per_token_ms: number;
    predicted_per_second: number;
  };
};

export type LlamaCompletionToken = {
  token: string;
};

export type TokenizeResult = {
  tokens: Int32Array;
};

export type EmbeddingResult = {
  embedding: Float32Array;
};

export interface LlamaContext {
  new (options: LlamaModelOptions): LlamaContext;
  getSystemInfo(): string;
  getModelInfo(): object;
  getFormattedChat(
    messages: ChatMessage[],
    chat_template?: string,
    params?: {
      jinja?: boolean;
      response_format?: CompletionResponseFormat;
      tools?: object;
      parallel_tool_calls?: object;
      tool_choice?: string;
    }
  ): object | string;
  completion(
    options: LlamaCompletionOptions,
    callback?: (token: LlamaCompletionToken) => void
  ): Promise<LlamaCompletionResult>;
  stopCompletion(): void;
  tokenize(text: string): Promise<TokenizeResult>;
  detokenize(tokens: number[]): Promise<string>;
  embedding(text: string): Promise<EmbeddingResult>;
  saveSession(path: string): Promise<void>;
  loadSession(path: string): Promise<void>;
  release(): Promise<void>;
  applyLoraAdapters(adapters: { path: string; scaled: number }[]): void;
  removeLoraAdapters(adapters: { path: string }[]): void;
  getLoadedLoraAdapters(): { path: string; scaled: number }[];
  // static
  loadModelInfo(path: string, skip: string[]): Promise<Object>;
}

export interface Module {
  LlamaContext: LlamaContext;
}

export type LibVariant = "default" | "vulkan" | "cuda";

const setupEnv = (variant?: string) => {
  const postfix = variant ? `-${variant}` : "";
  const binPath = path.resolve(
    __dirname,
    `../bin/${process.platform}${postfix}/${process.arch}/`
  );
  const systemPathEnv = process.env.PATH ?? process.env.Path ?? "";
  if (!systemPathEnv.includes(binPath)) {
    if (process.platform === "win32") {
      process.env.Path = `${binPath};${systemPathEnv}`;
    } else {
      process.env.PATH = `${binPath}:${systemPathEnv}`;
    }
  }
};

export const loadModule = async (variant?: LibVariant): Promise<Module> => {
  try {
    if (variant && variant !== "default") {
      setupEnv(variant);
      return (await import(
        `../bin/${process.platform}-${variant}/${process.arch}/llama-node.node`
      )) as Module;
    }
  } catch {} // ignore errors and try the common path
  setupEnv();
  return (await import(
    `../bin/${process.platform}/${process.arch}/llama-node.node`
  )) as Module;
};
