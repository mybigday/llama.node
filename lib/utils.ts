
import type {
  ChatMessage,
} from './binding'

export const MTMD_DEFAULT_MEDIA_MARKER = '<__media__>'

export const formatMediaChat = (messages: ChatMessage[] | undefined): {
  messages: ChatMessage[] | undefined
  has_media: boolean
  media_paths?: string[]
} => {
  if (!messages)
    return {
      messages,
      has_media: false,
    }
  const mediaPaths: string[] = []
  return {
    messages: messages.map((msg) => {
      if (Array.isArray(msg.content)) {
        const content = msg.content.map((part) => {
          // Handle multimodal content
          if (part.type === 'image_url') {
            let path = part.image_url?.url || ''
            mediaPaths.push(path)
            return {
              type: 'text',
              text: MTMD_DEFAULT_MEDIA_MARKER,
            }
          } else if (part.type === 'input_audio') {
            const { input_audio: audio } = part
            if (!audio) throw new Error('input_audio is required')

            const { format } = audio
            if (format != 'wav' && format != 'mp3') {
              throw new Error(`Unsupported audio format: ${format}`)
            }
            if (audio.url) {
              const path = audio.url.replace(/file:\/\//, '')
              mediaPaths.push(path)
            } else if (audio.data) {
              mediaPaths.push(audio.data)
            }
            return {
              type: 'text',
              text: MTMD_DEFAULT_MEDIA_MARKER,
            }
          }
          return part
        })

        return {
          ...msg,
          content,
        }
      }
      return msg
    }),
    has_media: mediaPaths.length > 0,
    media_paths: mediaPaths,
  }
}