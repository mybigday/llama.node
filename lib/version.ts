export const BUILD_NUMBER = '6096';
export const BUILD_COMMIT = 'fd1234cb';
