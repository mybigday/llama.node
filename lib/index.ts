import { loadModule, LlamaModelOptions } from './binding'
import type {
  Module,
  LlamaContext,
  LibVariant,
  ChatMessage,
  LlamaCompletionOptions,
  LlamaCompletionToken,
  LlamaCompletionResult,
  TokenizeResult,
  EmbeddingResult,
  RerankParams,
  RerankResult,
  CompletionResponseFormat,
  ModelInfo,
  JinjaFormattedChatResult,
  Tool,
  GGUFModelInfo,
} from './binding'
import { BUILD_NUMBER, BUILD_COMMIT } from './version'
import { LlamaParallelAPI } from './parallel'

export * from './binding'
export { LlamaParallelAPI }

export const MTMD_DEFAULT_MEDIA_MARKER = '<__media__>'

export interface LlamaModelOptionsExtended extends LlamaModelOptions {
  lib_variant?: LibVariant
}

const mods: { [key: string]: Module } = {}

const logListeners: Array<(level: string, text: string) => void> = []

const logCallback = (level: string, text: string) => {
  logListeners.forEach((listener) => listener(level, text))
}

let logEnabled = false

const refreshNativeLogSetup = () => {
  Object.entries(mods).forEach(([, mod]) => {
    mod.LlamaContext.toggleNativeLog(logEnabled, logCallback)
  })
}

export const toggleNativeLog = async (enable: boolean) => {
  logEnabled = enable
  refreshNativeLogSetup()
}

export function addNativeLogListener(
  listener: (level: string, text: string) => void,
): { remove: () => void } {
  logListeners.push(listener)
  return {
    remove: () => {
      logListeners.splice(logListeners.indexOf(listener), 1)
    },
  }
}

const getJsonSchema = (responseFormat?: CompletionResponseFormat) => {
  if (responseFormat?.type === 'json_schema') {
    return responseFormat.json_schema?.schema
  }
  if (responseFormat?.type === 'json_object') {
    return responseFormat.schema || {}
  }
  return null
}

export type FormattedChatResult = {
  type: 'jinja' | 'llama-chat'
  prompt: string
  has_media: boolean
  media_paths?: Array<string>
}

class LlamaContextWrapper {
  ctx: LlamaContext
  parallel: LlamaParallelAPI

  constructor(nativeCtx: LlamaContext) {
    this.ctx = nativeCtx
    this.parallel = new LlamaParallelAPI(nativeCtx)
  }

  getSystemInfo(): string {
    return this.ctx.getSystemInfo()
  }

  getModelInfo(): ModelInfo {
    return this.ctx.getModelInfo()
  }

  isJinjaSupported(): boolean {
    const { minja } = this.ctx.getModelInfo().chatTemplates
    return !!minja?.toolUse || !!minja?.default
  }

  isLlamaChatSupported(): boolean {
    return !!this.ctx.getModelInfo().chatTemplates.llamaChat
  }

  _formatMediaChat(messages: ChatMessage[] | undefined): {
    messages: ChatMessage[] | undefined
    has_media: boolean
    media_paths?: string[]
  } {
    if (!messages)
      return {
        messages,
        has_media: false,
      }
    const mediaPaths: string[] = []
    return {
      messages: messages.map((msg) => {
        if (Array.isArray(msg.content)) {
          const content = msg.content.map((part) => {
            // Handle multimodal content
            if (part.type === 'image_url') {
              let path = part.image_url?.url || ''
              mediaPaths.push(path)
              return {
                type: 'text',
                text: MTMD_DEFAULT_MEDIA_MARKER,
              }
            } else if (part.type === 'input_audio') {
              const { input_audio: audio } = part
              if (!audio) throw new Error('input_audio is required')

              const { format } = audio
              if (format != 'wav' && format != 'mp3') {
                throw new Error(`Unsupported audio format: ${format}`)
              }
              if (audio.url) {
                const path = audio.url.replace(/file:\/\//, '')
                mediaPaths.push(path)
              } else if (audio.data) {
                mediaPaths.push(audio.data)
              }
              return {
                type: 'text',
                text: MTMD_DEFAULT_MEDIA_MARKER,
              }
            }
            return part
          })

          return {
            ...msg,
            content,
          }
        }
        return msg
      }),
      has_media: mediaPaths.length > 0,
      media_paths: mediaPaths,
    }
  }

  getFormattedChat(
    messages: ChatMessage[],
    template?: string,
    params?: {
      jinja?: boolean
      response_format?: CompletionResponseFormat
      tools?: Tool[]
      parallel_tool_calls?: boolean
      tool_choice?: string
      enable_thinking?: boolean
      add_generation_prompt?: boolean
      now?: string | number
      chat_template_kwargs?: Record<string, string>
    },
  ): FormattedChatResult {
    const {
      messages: chat,
      has_media,
      media_paths,
    } = this._formatMediaChat(messages)

    const useJinja = this.isJinjaSupported() && params?.jinja
    let tmpl
    if (template) tmpl = template // Force replace if provided

    const result = this.ctx.getFormattedChat(chat!, tmpl, {
      jinja: useJinja,
      response_format: params?.response_format,
      tools: params?.tools,
      parallel_tool_calls: params?.parallel_tool_calls,
      tool_choice: params?.tool_choice,
      enable_thinking: params?.enable_thinking ?? true,
      add_generation_prompt: params?.add_generation_prompt,
      now: params?.now,
      chat_template_kwargs: params?.chat_template_kwargs
        ? Object.entries(params.chat_template_kwargs).reduce(
            (acc, [key, value]) => {
              acc[key] = JSON.stringify(value) // Each value is a stringified JSON object
              return acc
            },
            {} as Record<string, any>,
          )
        : undefined,
    })

    if (!useJinja) {
      return {
        type: 'llama-chat',
        prompt: result as string,
        has_media,
        media_paths,
      }
    }
    const jinjaResult = result as JinjaFormattedChatResult
    return {
      type: 'jinja',
      has_media,
      media_paths,
      ...jinjaResult,
    }
  }

  completion(
    options: LlamaCompletionOptions,
    callback?: (token: LlamaCompletionToken) => void,
  ): Promise<LlamaCompletionResult> {
    const { messages, media_paths = options.media_paths } =
      this._formatMediaChat(options.messages)
    return this.ctx.completion(
      {
        ...options,
        messages,
        media_paths: options.media_paths || media_paths,
      },
      callback || (() => {}),
    )
  }

  stopCompletion(): void {
    return this.ctx.stopCompletion()
  }

  tokenize(
    text: string,
    { media_paths }: { media_paths?: string[] } = {},
  ): Promise<TokenizeResult> {
    return this.ctx.tokenize(text, media_paths)
  }

  detokenize(tokens: number[]): Promise<string> {
    return this.ctx.detokenize(tokens)
  }

  embedding(text: string, params?: { embd_normalize?: number }): Promise<EmbeddingResult> {
    return this.ctx.embedding(text, params)
  }

  rerank(
    query: string,
    documents: string[],
    params?: RerankParams,
  ): Promise<Array<RerankResult & { document: string }>> {
    return this.ctx
      .rerank(query, documents, params)
      .then((results: RerankResult[]) => {
        // Sort by score descending and add document text for convenience
        return results
          .map((result: RerankResult) => ({
            ...result,
            document: documents[result.index],
          }))
          .sort(
            (
              a: RerankResult & { document: string },
              b: RerankResult & { document: string },
            ) => b.score - a.score,
          )
      })
  }

  saveSession(path: string): Promise<void> {
    return this.ctx.saveSession(path)
  }

  loadSession(path: string): Promise<void> {
    return this.ctx.loadSession(path)
  }

  release(): Promise<void> {
    return this.ctx.release()
  }

  applyLoraAdapters(adapters: { path: string; scaled: number }[]): void {
    return this.ctx.applyLoraAdapters(adapters)
  }

  removeLoraAdapters(): void {
    this.ctx.removeLoraAdapters()
  }

  getLoadedLoraAdapters(): { path: string; scaled: number }[] {
    return this.ctx.getLoadedLoraAdapters()
  }

  initMultimodal(options: { path: string; use_gpu?: boolean }): boolean {
    return this.ctx.initMultimodal(options)
  }

  isMultimodalEnabled(): boolean {
    return this.ctx.isMultimodalEnabled()
  }

  releaseMultimodal(): void {
    this.ctx.releaseMultimodal()
  }

  getMultimodalSupport(): {
    vision: boolean
    audio: boolean
  } {
    return this.ctx.getMultimodalSupport()
  }

  initVocoder(options: { path: string; n_batch?: number }): boolean {
    return this.ctx.initVocoder(options)
  }

  releaseVocoder(): void {
    this.ctx.releaseVocoder()
  }

  isVocoderEnabled(): boolean {
    return this.ctx.isVocoderEnabled()
  }

  getFormattedAudioCompletion(
    speaker: string | null,
    text: string,
  ): {
    prompt: string
    grammar?: string
  } {
    return this.ctx.getFormattedAudioCompletion(speaker, text)
  }

  getAudioCompletionGuideTokens(text: string): Int32Array {
    return this.ctx.getAudioCompletionGuideTokens(text)
  }

  decodeAudioTokens(tokens: number[] | Int32Array): Promise<Float32Array> {
    return this.ctx.decodeAudioTokens(tokens)
  }
}

export const loadModel = async (
  options: LlamaModelOptionsExtended,
): Promise<LlamaContextWrapper> => {
  const variant = options.lib_variant ?? 'default'
  mods[variant] ??= await loadModule(options.lib_variant)
  refreshNativeLogSetup()

  const nativeCtx = new mods[variant].LlamaContext(options)
  return new LlamaContextWrapper(nativeCtx)
}

export const initLlama = loadModule

const modelInfoSkip = [
  // Large fields
  'tokenizer.ggml.tokens',
  'tokenizer.ggml.token_type',
  'tokenizer.ggml.merges',
  'tokenizer.ggml.scores',
]

export const loadLlamaModelInfo = async (
  path: string,
): Promise<GGUFModelInfo> => {
  const variant = 'default'
  mods[variant] ??= await loadModule(variant)
  refreshNativeLogSetup()
  return mods[variant].LlamaContext.loadModelInfo(path, modelInfoSkip)
}

export const BuildInfo = {
  number: BUILD_NUMBER,
  commit: BUILD_COMMIT,
}
