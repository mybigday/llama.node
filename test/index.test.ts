import path from 'path'
import waitForExpect from 'wait-for-expect'
import { loadModel } from '../lib'

it('work fine', async () => {
  let tokens = ''
  const model = await loadModel({ model: path.resolve(__dirname, './tiny-random-llama.gguf') })
  const info = model.getModelInfo()
  expect(info).toMatchSnapshot('model info')
  const result = await model.completion({
    prompt: 'My name is Merve and my favorite',
    n_samples: 1,
    temperature: 0,
    max_length: 100,
    n_predict: 10,
    seed: 0,
  }, (data) => {
    expect(data).toMatchObject({ token: expect.any(String) })
    tokens += data.token
  })
  expect({
    ...result,
    timings: `Timings: (${Object.keys(result.timings).length}) keys`
  }).toMatchSnapshot()
  await waitForExpect(() => {
    expect(tokens).toBe(result.text)
  })
  await model.saveSession(path.resolve(__dirname, './tmp.sess'))
  await model.loadSession(path.resolve(__dirname, './tmp.sess'))
  await model.release()
})

it('work fine with vocab_only', async () => {
  const model = await loadModel({ model: path.resolve(__dirname, './tiny-random-llama.gguf'), vocab_only: true })
  expect(model.getModelInfo()).toMatchSnapshot('model info')
  expect(await model.tokenize('Once upon a time')).toMatchSnapshot('tokenize')
  expect(await model.completion({ prompt: 'Once upon a time' })).toMatchSnapshot('empty result')
})

it('tokeneize', async () => {
  const model = await loadModel({ model: path.resolve(__dirname, './tiny-random-llama.gguf') })
  {
    const result = await model.tokenize('Once upon a time')
    expect(result).toMatchSnapshot()
  }
  {
    const result = await model.detokenize([123, 123, 123])
    expect(result).toMatchSnapshot()
  }
  {
    const result = model.getFormattedChat([
      { role: 'user', text: 'Hello' },
      { role: 'bot', text: 'Hi' },
    ])
    expect(result).toMatchSnapshot()
  }
  await model.release()
})

it('embedding', async () => {
  const model = await loadModel({
    model: path.resolve(__dirname, './bge-small-en.gguf'),
    embedding: true,
    n_gpu_layers: 0,
  })
  const result = await model.embedding('Once upon a time')
  expect(result).toMatchSnapshot()
  await model.release()
})
