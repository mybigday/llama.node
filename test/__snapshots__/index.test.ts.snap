// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic completion 1`] = `
{
  "chat_format": 0,
  "content": "<removed>",
  "context_full": false,
  "interrupted": false,
  "stopped_eos": false,
  "stopped_limited": true,
  "stopped_words": false,
  "stopping_word": "",
  "text": "wallwallwallwallwallwallwallwallwallwall",
  "timings": "Timings: (8) keys",
  "tokens_evaluated": 18,
  "tokens_predicted": 9,
  "truncated": false,
}
`;

exports[`basic completion: model info 1`] = `
{
  "chatTemplates": {
    "llamaChat": false,
    "minja": {
      "default": false,
      "defaultCaps": {
        "parallelToolCalls": false,
        "systemRole": true,
        "toolCallId": false,
        "toolCalls": false,
        "toolResponses": false,
        "tools": false,
      },
      "toolUse": false,
    },
  },
  "desc": "llama ?B Q4_0",
  "isChatTemplateSupported": false,
  "metadata": {
    "general.architecture": "llama",
    "general.basename": "Llama-3.2",
    "general.file_type": "2",
    "general.finetune": "1b",
    "general.name": "Llama 3.2 1B",
    "general.organization": "Meta Llama",
    "general.quantization_version": "2",
    "general.size_label": "73M",
    "general.type": "model",
    "llama.attention.head_count": "8",
    "llama.attention.head_count_kv": "2",
    "llama.attention.key_length": "64",
    "llama.attention.layer_norm_rms_epsilon": "0.000010",
    "llama.attention.value_length": "64",
    "llama.block_count": "2",
    "llama.context_length": "131072",
    "llama.embedding_length": "512",
    "llama.feed_forward_length": "2048",
    "llama.rope.dimension_count": "64",
    "llama.rope.freq_base": "500000.000000",
    "llama.vocab_size": "128256",
    "tokenizer.ggml.bos_token_id": "128000",
    "tokenizer.ggml.eos_token_id": "128001",
    "tokenizer.ggml.model": "gpt2",
    "tokenizer.ggml.pre": "llama-bpe",
  },
  "nEmbd": 512,
  "nParams": 73271840,
  "size": 58154112,
}
`;

exports[`completion with response_format: json_object 1`] = `
{
  "chat_format": 0,
  "content": "<removed>",
  "context_full": false,
  "interrupted": false,
  "stopped_eos": true,
  "stopped_limited": false,
  "stopped_words": false,
  "stopping_word": "",
  "text": "{}
		              			",
  "timings": "Timings: (8) keys",
  "tokens_evaluated": 12,
  "tokens_predicted": 3,
  "truncated": false,
}
`;

exports[`completion with response_format: json_schema 1`] = `
{
  "chat_format": 0,
  "content": "<removed>",
  "context_full": false,
  "interrupted": false,
  "stopped_eos": false,
  "stopped_limited": true,
  "stopped_words": false,
  "stopping_word": "",
  "text": "{
 				 				 				 				",
  "timings": "Timings: (8) keys",
  "tokens_evaluated": 13,
  "tokens_predicted": 4,
  "truncated": false,
}
`;

exports[`completion with t5-like model 1`] = `
{
  "chat_format": 0,
  "content": "<removed>",
  "context_full": false,
  "interrupted": false,
  "stopped_eos": true,
  "stopped_limited": false,
  "stopped_words": false,
  "stopping_word": "",
  "text": " How old sind er?",
  "timings": "Timings: (8) keys",
  "tokens_evaluated": 17,
  "tokens_predicted": 6,
  "truncated": false,
}
`;

exports[`completion with tools 1`] = `
{
  "additional_stops": [],
  "chat_format": 9,
  "grammar": "any-tool-call ::= ( calc-call ) space
calc-args ::= "{" space  (calc-args-expression-kv )? "}" space
calc-args-expression-kv ::= "\\"expression\\"" space ":" space string
calc-call ::= "{" space calc-call-name-kv "," space calc-call-arguments-kv "}" space
calc-call-arguments ::= "{" space  (calc-call-arguments-expression-kv )? "}" space
calc-call-arguments-expression-kv ::= "\\"expression\\"" space ":" space string
calc-call-arguments-kv ::= "\\"arguments\\"" space ":" space calc-call-arguments
calc-call-name ::= "\\"calc\\"" space
calc-call-name-kv ::= "\\"name\\"" space ":" space calc-call-name
calc-function-tag ::= "<function" ( "=calc" | " name=\\"calc\\"" ) ">" space calc-args "</function>" space
char ::= [^"\\\\\\x7F\\x00-\\x1F] | [\\\\] (["\\\\bfnrt] | "u" [0-9a-fA-F]{4})
root ::= tool-call
space ::= | " " | "\\n"{1,2} [ \\t]{0,20}
string ::= "\\"" char* "\\"" space
tool-call ::= calc-function-tag | wrappable-tool-call | ( "\`\`\`\\n" | "\`\`\`json\\n" | "\`\`\`xml\\n" ) space wrappable-tool-call space "\`\`\`" space 
wrappable-tool-call ::= ( any-tool-call | "<tool_call>" space any-tool-call "</tool_call>" | "<function_call>" space any-tool-call "</function_call>" | "<response>"  space any-tool-call "</response>" | "<tools>"     space any-tool-call "</tools>" | "<json>"      space any-tool-call "</json>" | "<xml>"      space any-tool-call "</xml>" | "<JSON>"      space any-tool-call "</JSON>" ) space
",
  "grammar_triggers": [
    {
      "token": -1,
      "type": 1,
      "value": "<function=calc>",
    },
    {
      "token": -1,
      "type": 2,
      "value": "<function\\s+name\\s*=\\s*"calc"",
    },
    {
      "token": -1,
      "type": 3,
      "value": "(?:<think>[\\s\\S]*?</think>\\s*)?(\\s*(?:<tool_call>|<function|(?:\`\`\`(?:json|xml)?
\\s*)?(?:<function_call>|<tools>|<xml><json>|<response>)?\\s*\\{\\s*"name"\\s*:\\s*"(?:calc)"))[\\s\\S]*",
    },
  ],
  "grammea_lazy": true,
  "has_media": false,
  "media_paths": [],
  "preserved_tokens": [
    "<think>",
    "</think>",
    "<tool_call>",
    "</tool_call>",
    "<function",
    "<tools>",
    "</tools>",
    "<response>",
    "</response>",
    "<function_call>",
    "</function_call>",
    "<json>",
    "</json>",
    "<JSON>",
    "</JSON>",
    "\`\`\`",
    "\`\`\`json",
    "\`\`\`xml",
  ],
  "prompt": "<|im_start|>system
# Tools

You may call one or more functions to assist with the user query.

You are provided with function signatures within <tools></tools> XML tags:
<tools>
{"type": "function", "function": {"name": "calc", "description": "Calculates the result of a math expression.", "parameters": {"type": "object", "properties": {"expression": {"type": "string", "description": "The math expression to evaluate."}}}}}
</tools>

For each function call, return a json object with function name and arguments within <tool_call></tool_call> XML tags:
<tool_call>
{"name": <function-name>, "arguments": <args-json-object>}
</tool_call><|im_end|>
<|im_start|>user
What is the sum of 1 and 2?<|im_end|>
<|im_start|>assistant
",
  "thinking_forced_open": false,
  "type": "jinja",
}
`;

exports[`loadModelInfo 1`] = `
{
  "alignment": 32,
  "data_offset": 7819680,
  "general.architecture": "llama",
  "general.basename": "Llama-3.2",
  "general.file_type": "2",
  "general.finetune": "1b",
  "general.name": "Llama 3.2 1B",
  "general.organization": "Meta Llama",
  "general.quantization_version": "2",
  "general.size_label": "73M",
  "general.type": "model",
  "llama.attention.head_count": "8",
  "llama.attention.head_count_kv": "2",
  "llama.attention.key_length": "64",
  "llama.attention.layer_norm_rms_epsilon": "0.000010",
  "llama.attention.value_length": "64",
  "llama.block_count": "2",
  "llama.context_length": "131072",
  "llama.embedding_length": "512",
  "llama.feed_forward_length": "2048",
  "llama.rope.dimension_count": "64",
  "llama.rope.freq_base": "500000.000000",
  "llama.vocab_size": "128256",
  "tokenizer.ggml.bos_token_id": "128000",
  "tokenizer.ggml.eos_token_id": "128001",
  "tokenizer.ggml.model": "gpt2",
  "tokenizer.ggml.pre": "llama-bpe",
  "version": 3,
}
`;

exports[`tokeneize & detokenize & getFormattedChat 1`] = `
{
  "bitmap_hashes": [],
  "chunk_pos": [],
  "chunk_pos_media": [],
  "has_media": false,
  "tokens": Int32Array [
    12805,
    5304,
    264,
    892,
  ],
}
`;

exports[`tokeneize & detokenize & getFormattedChat 2`] = `"���"`;

exports[`tokeneize & detokenize & getFormattedChat 3`] = `
{
  "has_media": false,
  "media_paths": [],
  "prompt": "<|im_start|>user
Hello<|im_end|>
<|im_start|>bot
Hi<|im_end|>
<|im_start|>assistant
",
  "type": "llama-chat",
}
`;

exports[`tokeneize & detokenize & getFormattedChat 4`] = `
{
  "has_media": false,
  "media_paths": [],
  "prompt": "<|im_start|>user
Hello<|im_end|>
<|im_start|>bot
Hi<|im_end|>
<|im_start|>assistant
",
  "type": "llama-chat",
}
`;

exports[`works fine with vocab_only: empty result 1`] = `
{
  "chat_format": 0,
  "context_full": false,
  "interrupted": false,
  "stopped_eos": false,
  "stopped_limited": true,
  "stopped_words": false,
  "stopping_word": "",
  "text": "",
  "timings": {
    "predicted_ms": 0,
    "predicted_n": 1,
    "predicted_per_second": Infinity,
    "predicted_per_token_ms": 0,
    "prompt_ms": 0,
    "prompt_n": 1,
    "prompt_per_second": Infinity,
    "prompt_per_token_ms": 0,
  },
  "tokens_evaluated": 0,
  "tokens_predicted": 0,
  "truncated": false,
}
`;

exports[`works fine with vocab_only: model info 1`] = `
{
  "chatTemplates": {
    "llamaChat": false,
    "minja": {
      "default": false,
      "defaultCaps": {
        "parallelToolCalls": false,
        "systemRole": true,
        "toolCallId": false,
        "toolCalls": false,
        "toolResponses": false,
        "tools": false,
      },
      "toolUse": false,
    },
  },
  "desc": "",
  "isChatTemplateSupported": false,
  "metadata": {
    "general.architecture": "llama",
    "general.basename": "Llama-3.2",
    "general.file_type": "2",
    "general.finetune": "1b",
    "general.name": "Llama 3.2 1B",
    "general.organization": "Meta Llama",
    "general.quantization_version": "2",
    "general.size_label": "73M",
    "general.type": "model",
    "llama.attention.head_count": "8",
    "llama.attention.head_count_kv": "2",
    "llama.attention.key_length": "64",
    "llama.attention.layer_norm_rms_epsilon": "0.000010",
    "llama.attention.value_length": "64",
    "llama.block_count": "2",
    "llama.context_length": "131072",
    "llama.embedding_length": "512",
    "llama.feed_forward_length": "2048",
    "llama.rope.dimension_count": "64",
    "llama.rope.freq_base": "500000.000000",
    "llama.vocab_size": "128256",
    "tokenizer.ggml.bos_token_id": "128000",
    "tokenizer.ggml.eos_token_id": "128001",
    "tokenizer.ggml.model": "gpt2",
    "tokenizer.ggml.pre": "llama-bpe",
  },
  "nEmbd": 0,
  "nParams": 73271840,
  "size": 58154112,
}
`;

exports[`works fine with vocab_only: tokenize 1`] = `
{
  "bitmap_hashes": [],
  "chunk_pos": [],
  "chunk_pos_media": [],
  "has_media": false,
  "tokens": Int32Array [
    12805,
    5304,
    264,
    892,
  ],
}
`;
